### PR TITLE
Integrate templates into LaTeX kernel

### PR DESCRIPTION
**Pull request by josephwright**
# Internal housekeeping

## Status of pull request

- Feedback wanted 
- Under development
- [x] Ready to merge

## Checklist of required changes before merge will be approved
- [x] Test file(s) added
- [x] Version and date string updated in changed source files
- [ ] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [ ] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

# Notes

This includes the changes
- Renaming object to template types
- Dropping restricted templates
- Providing template copying

### CODE REVIEW

File: base/doc/ltnews38.tex [hunk 1]
```diff
@@ -148,6 +148,14 @@ \section{Sockets and plugs}
 
 For documentation see \texttt{texdoc ltsockets-doc} for now.
 
+\section{Providing \pkg{xtemplate} in the format}
+
+In \LaTeX{} News~32, we decribed the move of one long-term experimental
```
Comment (mrpiggi) [inline base/doc/ltnews38.tex:153]: typo: decribed > described

Comment (josephwright) [reply base/doc/ltnews38.tex:153]: Yeah - I have to actually write this yet :)

**Comment by car222222** [comment]
The ltnews38 entry confuses me.  

Are we sure that this can simply be integrated in as here, without at least the full, updated documentation. 

Also, are we sure that all the currently suggested extensions (that have been extensively discussed elsewhere) can be added smoothly to this version?

**Comment by josephwright** [comment]
> The ltnews38 entry confuses me.

Er, that part is not finished: it's always the most tricky aspect ;)

> Are we sure that this can simply be integrated in as here, without at least the full, updated documentation.

Definitely not! I need to write a proper entry and I assume we will move the documentation to `clsguide` from `lttemplate` - but I first wanted to be sure that the overall scheme is OK.

> Also, are we sure that all the currently suggested extensions (that have been extensively discussed elsewhere) can be added smoothly to this version?

Do you mean inheritance?

**Comment by car222222** [comment]
I hope this is not too difficult -- Please can someone provide a complete list of all the changes made compared to last week's version of xtemplate.  

This would make reviewing much easier. 

It currently lists some, but it implies there are others: "This includes the changes . . . ".

**Comment by car222222** [comment]
"Do you mean inheritance?"  Maybe! 

Are you aware of the extensive recent discussions covering this and many other related matters? 

It is good to know that there is still a lot of work to do on this integration.  But for that reason, amongst others, it has been suggested that this need (and probably should) not proceed right now.

**Comment by car222222** [comment]
"move the documentation"

What about the extensive documentation about actually using the stuff in practice?  This dates from (roughly) 2013 and needs some updating and careful checking.  It then needs a home! 

I would like to see a full plan for everything that needs to be done to make this stuff usefully available with full documentation, including many examples.  And probably some kind of agreed time-line and division of work-load for the whole process.

**Comment by josephwright** [comment]
> What about the extensive documentation about actually using the stuff in practice? This dates from (roughly) 2013 and needs some updating and careful checking. It then needs a home!

Well it's really a lot older: parts go back to `template2` at least. But it's as far as it's all correct. In terms of examples, isn't that what `xfrac` and (to a lesser extend) `xgalley` are for? We could take some examples from those and/or the more recent work Frank has done for block elements and explain. But I'm not sure that would fit into `clsguide`: feels like it would be too long.

> I would like to see a full plan for everything that needs to be done to make this stuff usefully available with full documentation, including many examples. And probably some kind of agreed time-line and division of work-load for the whole process.

Given that there are 'in the wild' users of `xtemplate`, I'm hoping it's not too opaque to understand. But we can certainly extend the docs.

**Comment by josephwright** [comment]
> Are you aware of the extensive recent discussions covering this and many other related matters?

I suspect I must have missed some things here: could you add to the issue for reference?

**Comment by josephwright** [comment]
> I hope this is not too difficult -- Please can someone provide a complete list of all the changes made compared to last week's version of xtemplate.

At the *interface* level, the changes are

- Rename object types to template types
- Remove support for restricted templates
- Remove support for collections (was already deprecated in `xtemplates`)

The rest of the changes are within the code, for example

- Renames for template type, including in the messages
- Removal of arguments for collection information
- Tidying of `msg` usage (e.g. `\msg_info:nnVV` in place of `\msg_info:nnxx`), including removing some completely unused arguments
- Standardisation of passing of type/template pair: as two arguments not one formatting `{ <type> / <template> }` (older version was inconsistet)
- Improvements to `l3doc` mark-up
- Kernel-specfiic integration, e.g. adding `template` to the `LaTeX` namespace (so messages are '`LaTeX info`' not '`Package template info`'
- Corrections to some errors (I picked up one in how a message is constructed)
- More consistent use of `\cs_generate_variant:Nn` rather than `\exp_args:N...`

**Comment by car222222** [comment]
So I must (simply:-) read the new version, trying to link it correctly to my extensive annotations to the old version.  Sigh!  "I may be some time".

At least it is shorter now!

Does it contain anything about possible future extensions, eg for the diffcoeff stuff?

**Comment by car222222** [comment]
If I had known that this was suddenly springing into life again, despite the very clear recent movement towards delaying it, I would have sent you all my detailed comments on xtemplate.

Now it seems that I have to start from scratch in transcribing them to the suddenly existing (without warning) new version . . . grrr!

**Comment by car222222** [comment]
> Are you aware of the extensive recent discussions covering this and many other related matters?

> I suspect I must have missed some things here: could you add to the issue for reference?

I can do this at some stage, after I have collated these together with some of my personal, unpublicised so far, ideas for compatible changes, and possible future extensions (to both the concepts and the provided facilities).

**Comment by car222222** [comment]
PS: Thanks for the list of changes:-).

**Comment by josephwright** [comment]
> So I must (simply:-) read the new version, trying to link it correctly to my extensive annotations to the old version. Sigh! "I may be some time".

Assuming you are looking at the documentation part, then the text is very much the same as before. Other than the rename and the targetted removals, I've not done anything there.

> Does it contain anything about possible future extensions, eg for the diffcoeff stuff?

No. I know there are different approaches here, but I tend toward logging and issue then (perhaps) creating a branch rather than adding anything speculative to sources.

**Comment by josephwright** [comment]
> Now it seems that I have to start from scratch in transcribing them to the suddenly existing (without warning) new version . . . grrr!

Comments on `xtemplate` would be fine: as I know what's different, it makes more sense for me to transcribe than for you to do it.

**Comment by josephwright** [comment]
I've converted to a draft as there is clearly more to discuss here.

**Comment by josephwright** [comment]
I've rebased after the kernel release, so we are now back to 'ready to go' at this side: I need to work on the L3 aspects

**Comment by car222222** [comment]
Meaning what exactly?

That it is now time to make necessary (before release) corrections and (maybe) some enhancements/flexibilities.

**Comment by car222222** [comment]
Meaning what exactly?

That it is now time to make necessary (before release) corrections and (maybe) some enhancements/flexibilities.

**Comment by josephwright** [comment]
> Meaning what exactly?
> 
> That it is now time to make necessary (before release) corrections and (maybe) some enhancements/flexibilities.

@car222222 Meaning that I've updated so that the code in the branch makes sense relative to the 2023-11-01 LaTeX release and so there won't be distracting comments about e.g. News being aimed at the wrong file. I've left the PR as a draft as I know there are things you feel need further work - I'm still though not clear on exactly what they are.

**Comment by car222222** [comment]
So does that mean that the next stage is to discuss any changes that need to be made to the code?

Should such detailed discussions take place under this PR?  

Or is there a preference amongst some of us for a different forum? 

I have some such "ready to go" (precise ideas, not actual code changes!).

**Comment by car222222** [comment]
That should have been "almost ready to go" :-).

**Comment by josephwright** [comment]
@car222222  I think comments here would be sensible, unless we end up needing several parallel threads (for which GitHub Issues work poorly)

**Comment by FrankMittelbach** [comment]
> @car222222 I think comments here would be sensible, unless we end up needing several parallel threads (for which GitHub Issues work poorly)

in other words, if it is about adjustments/suggestions directly related to existing code, so that they can be resolved in situ, the here as review to the code. But if it is about bigger alternative suggestions, then probably better first on the team list and if through that @josephwright makes updates to this pr, then resolving any new issues/suggestions concerning the updated code again next to that code here.

**Comment by josephwright** [comment]
As well as thee renaming, I'd like some ideas on how much to put in LaTeX News here.

**Comment by car222222** [comment]
Is LaTeX news relevant at this stage?

Let's first finish getting this done and put it into the kernel; and then do the necessary rewrite of documentation in the file, and maybe elsewhere.   

But if you need something right now, just say:

"The templates concept has been updated and moved to the ?? kernel, in the file <name?>.
Read more about it here . . . , and here . . . ."

**Comment by josephwright** [comment]
> And maybe also produce somewhere something quite detailed and technical, to send to "known template developers" (and maybe also to other "friendly developers" who should be encouraged to use them), specifically asking for their input on our actions and ideas.

Indeed: I was thinking much the same. A quick survey of TeX Live shows there are currently six authors/author groups we'd need to contact to get everyone using temples: package names are
- `acro` (etc.)
- `diffcoeff`
- `fduthesis`
- `njuthesis`
- `nwafuthesis`
- `xeCJK`

**Pull request merged**